lib/puppet/provider/ssh_authorized_key/parsed.rb

```diff
@@ -56,16 +74,24 @@ def flush
     # so calling it here suppresses the later attempt by our superclass's flush method.
     self.class.backup_target(target)
 
-    Puppet::Util::SUIDManager.asuser(@resource.should(:user)) do
-      unless Puppet::FileSystem.exist?(dir = File.dirname(target))
-        Puppet.debug "Creating #{dir} as #{@resource.should(:user)}"
-        Dir.mkdir(dir, dir_perm)
-      end
-
+    # if the target path is trusted, create the file as root
+    if trusted_path
       super
-
-      File.chmod(file_perm, target)
+      user = @resource.should(:user)
+      gid = Etc.getpwnam(user).gid
+      FileUtils.chown(user, gid, target)
+    # otherwise create the file as the specified user
+    else
+      Puppet::Util::SUIDManager.asuser(@resource.should(:user)) do
```

Comment:
The type claims that either `user` or `target` are required. But the flush method will raise if `@resource.should(:user)` is nil, meaning it wasn't specified in the manifest (so `user` is not currently optional)`.

I think we should change the flush method so that if `user` is specified, then we always switch uid and write the file as that user (as the module did before). If `user` is not specified and path is trusted, then write the file to the specified `target` as the current user. If the `user` is not specified and the path is not trusted, then fail.

Reply:
That makes the most sense. I will implement that.

Reply:
> If user is not specified and path is trusted, then write the file to the specified target as the current user.

Actually I don't think we can do this, because if `user` is not present we don't know who to `chown` the file to.

Reply:
Ok makes sense, then `:user` should be always required.

* Please delete https://github.com/puppetlabs/puppetlabs-sshkeys_core/blob/master/lib/puppet/type/ssh_authorized_key.rb#L139-L140
* Update https://github.com/puppetlabs/puppetlabs-sshkeys_core/blob/master/lib/puppet/type/ssh_authorized_key.rb#L145-L146 to say the `user` attribute is required
* We need to be very CAREFUL about chmod'ing files as root due to time-of-check-time-of-use race conditions. See a previous CVE in this module: https://groups.google.com/forum/#!topic/puppet-announce/keO0bSMoocs. So for this PR, please do 2 things:
  * Add a comment about race conditions when modifying file permissions as root
  * Move the `File.chmod(file_perm, target)` call inside the block (when running with reduced privileges as the `user`)
  * Modify the `trusted_path` case so that we open a file handle to the `target` and we call `File#chown` and `File#chmod` instance methods on the file handle.
* I don't want the module to silently prefer "writing as root" if `trusted_path` happens to return true. I would prefer that the manifest author opt-into writing as root due to possible security implications. Let's create a new parameter, something like `drop_privileges` which defaults to `true`. The logic should then be:
  * If drop_privileges, write to target (or default location)
  * If not drop_privileges, and trusted path, write as root
  * If not drop_privileges, and not trusted path, raise
* Note due to https://tickets.puppetlabs.com/browse/PUP-8442 the parameter can't default to false.

Reply:
I tried to make `:user` mandatory, but it seems to fudge up cases when using the `user` resource with the `purge_ssh_keys` attribute:

```puppet
user { 'nick':              
  ensure         => present,
  purge_ssh_keys => true,   
}
```

From what I understand, if you create an `ssh_authorized_key` resource without passing the `target` parameter, you can then execute the above manifest to purge the keys.
The manifest creates an `ssh_authorized_key` resource with `:user` unset and `target => '/home/nick/.ssh/authorized_keys'`. It seems to infer the username from the user resource.

That's why only one of user and target were mandatory when validating. That's also why `@resource.should(:user)` is checked again in the `flush` method. Gnarly.

Reply:
Based on the comment above I don't think we should change https://github.com/puppetlabs/puppetlabs-sshkeys_core/blob/master/lib/puppet/type/ssh_authorized_key.rb#L139-L140

Reply:
I think I addressed most of the issues, let me know if I missed something.